core/trino-main/src/main/java/io/trino/operator/aggregation/partial/PartialAggregationOutputProcessor.java

```diff
@@ -0,0 +1,120 @@
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static io.trino.spi.type.BooleanType.BOOLEAN;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Creates final output of the partial aggregation step based on either aggregated or raw input.
+ * <p>
+ * Partial aggregation step output has additional channels used to send
+ * raw input data when the partial aggregation is disabled.
+ * Those channels need to be present in the output, albeit empty, even if the
+ * partial aggregation is enabled.
+ * <p>
+ * The additional channels are:
+ * - mask channels used by any aggregation
+ * - every aggregation input channels that are not already in hash channels
+ * - a boolean rawInputMask channel that contains decision which input, aggregated or raw, should be used for a given position.
+ */
+public class PartialAggregationOutputProcessor
+{
+    private final List<GroupedAggregator> groupedAggregators;
+    private final List<Type> aggregationInputTypes;
+    private final int[] hashChannels;
+    private final int[] aggregationInputChannels;
+    private final int[] maskBlockChannels;
+
+    public static PartialAggregationOutputProcessor create(List<Integer> groupByChannels, Optional<Integer> hashChannel, List<AggregatorFactory> aggregatorFactories, List<? extends Type> aggregationInputTypes, List<Integer> aggregationInputChannels)
+    {
+        return new PartialAggregationOutputProcessor(groupByChannels, hashChannel, aggregatorFactories, aggregationInputTypes, aggregationInputChannels, ImmutableList.copyOf(aggregatorFactories.stream().map(AggregatorFactory::getMaskChannel).flatMapToInt(OptionalInt::stream).boxed().collect(toImmutableSet())));
+    }
+
+    private PartialAggregationOutputProcessor(List<Integer> groupByChannels, Optional<Integer> inputHashChannel, List<AggregatorFactory> aggregatorFactories, List<? extends Type> aggregationInputTypes, List<Integer> aggregationInputChannels, List<Integer> maskBlockChannels)
+    {
+        this.aggregationInputTypes = ImmutableList.copyOf(aggregationInputTypes);
+
+        this.groupedAggregators = requireNonNull(aggregatorFactories, "aggregatorFactories is null").stream().map(AggregatorFactory::createGroupedAggregator).collect(toImmutableList());
+        this.aggregationInputChannels = Ints.toArray(aggregationInputChannels);
+        this.maskBlockChannels = Ints.toArray(maskBlockChannels);
+        this.hashChannels = new int[groupByChannels.size() + (inputHashChannel.isPresent() ? 1 : 0)];
+        for (int i = 0; i < groupByChannels.size(); i++) {
+            hashChannels[i] = groupByChannels.get(i);
+        }
+        inputHashChannel.ifPresent(channelIndex -> hashChannels[groupByChannels.size()] = channelIndex);
+    }
+
+    public Page processAggregatedPage(Page page)
+    {
+        Block[] finalPage = new Block[page.getChannelCount() + maskBlockChannels.length + aggregationInputTypes.size() + 1];
```

Comment:
extract number of blocks as final variable. It should be same for `processAggregatedPage` and `processRawInputPage`